### PR TITLE
memdump: dump when program enters new anonymous VAD

### PR DESCRIPTION
**Pull request by icedevml**
New optional flag `--dump-exec`, dump VAD when program starts executing anonymous (not file-backed) VAD. This is a new pretty effective general purpose heuristic.

Huge part of this was taken from #1000, the IPT plugin will be made much thinner after this is merged.

### CODE REVIEW

**Comment by tklengyel** [comment]
@drakvuf-jenkins This is OK to test

File: src/plugins/memdump/memdump.cpp [hunk 1]
```diff
@@ -873,6 +874,173 @@ static event_response_t set_information_thread_hook_cb(drakvuf_t drakvuf, drakvu
...
+};
+
+struct access_fault_result_t: public call_result_t
+{
+    access_fault_result_t() : call_result_t(), fault_va() {}
+
+    addr_t fault_va;
+};
+
+std::set<std::pair<addr_t, addr_t>> known_vads;
```
Comment (kscieslinski) [inline src/plugins/memdump/memdump.cpp:890]: Does this has to be sorted? If no, then how about using std::unordered_set?

Comment (BonusPlay) [reply src/plugins/memdump/memdump.cpp:890]: hint `std::set<std::pair<>>` is just `std::map` :)
(even better `std::unordered_map).

**Comment by icedevml** [comment]
@drakvuf-jenkins Retest this please.

**Comment by icedevml** [comment]
@tklengyel looks like there is some issue with windows 10
https://ci.drakvuf.com/job/DRAKVUF-windows10/44/console

File: src/plugins/memdump/memdump.cpp (hunk 1 above)
Comment (icedevml) [reply src/plugins/memdump/memdump.cpp]: I will include this

Comment (BonusPlay) [reply src/plugins/memdump/memdump.cpp]: After internal discussion we came to conclusion that `std::unordered_set<std::pair>` is correct here

File: src/plugins/memdump/memdump.cpp [hunk 2]
```diff
@@ -873,6 +874,167 @@ static event_response_t set_information_thread_hook_cb(drakvuf_t drakvuf, drakvu
...
+                    }
+                }
+            }
+        }
+    }
+
+    if (!dll_name_str)
+        dll_name_str = missing_dll_name;
+
+    jsonfmt::print("execframe", drakvuf, info,
```
Comment (icedevml) [inline src/plugins/memdump/memdump.cpp:945]: :thinking:

File: src/plugins/memdump/memdump.cpp [hunk 3]
```diff
@@ -873,6 +874,167 @@ static event_response_t set_information_thread_hook_cb(drakvuf_t drakvuf, drakvu
     return VMI_EVENT_RESPONSE_NONE;
 }
 
+struct exec_fault_data
+{
+    memdump* plugin;
+    addr_t rip;
```
Comment (Id3aFly) [inline src/plugins/memdump/memdump.cpp:880]: Why is this variable named rip? I understand it as the starting virtual address of the page in which some execution will fault in the future?

Comment (icedevml) [reply src/plugins/memdump/memdump.cpp:880]: yes

Comment (icedevml) [reply src/plugins/memdump/memdump.cpp:880]: I will modify it to some better name like page_va

**Comment by icedevml** [comment]
@drakvuf-jenkins Retest this please.

File: src/plugins/memdump/memdump.cpp (hunk 3 above)
Comment (Id3aFly) [reply src/plugins/memdump/memdump.cpp:880]: You could do this at the ipt plugin as well

**Comment by icedevml** [comment]
I think this feature can be integrated into Hyperbee plugin instead, it would be just a few LOCs modification. Abandoning for now and I will probably make a PR to Hyperbee.

**Pull request closed without merging**